### DISCUSSION TITLE
Saving incomplete form responses to LocalStorage and Pre-populating form

### DISCUSSION THREAD

**Discussion by yuvi100** [ditdot-dev/vue-flow-form#163] [Q&A] [answered]
Hi there,

Just wondering if anyone could help me out in saving incomplete form data to local storage and pre-populating the form based on those local storage values (if they exist - otherwise populate answers as 'blank').

Any help would be appreciated!

Thanks

**Comment by spinn** [comment, marked as answer]
Hi,

you could use the `onAnswer` or `onStep` events to catch when the user answers a question and then save your whole `questions` array to localStorage (or only the answers so as to not use too much storage space), and then when constructing the array, set the `answer` of the `QuestionModel` to your saved answer. This field is currently undocumented but I've opened issue #165 to make sure we document this in the future.

If you're setting the questions as components, simply set their v-model to whatever value you have saved in localStorage.

Example:
```vue
<template>
  <flow-form
    ref="flowform"
    v-on:step="onStep"
    v-bind:questions="questions"
  />
</template>

<script>
  export default {
    name: 'example',

    components: {
      FlowForm
    },

    data() {
      questions: [
        new QuestionModel({
          id: 'first_name',
          tagline: 'Hi! Welcome to our demo survey 😊',
          title: 'What is your first name?',
          type: QuestionType.Text,
          required: true,
          placeholder: 'Start typing here...',
          answer: localStorage.getItem('first_name')
        }),
        ...
    },

    onStep() {
      localStorage.setItem('first_name', this.questions[0].answer)
    },

    ...
  }
</script>
```

**Reply by yuvi100** [reply]
Thanks Dennis!

I got it to work, i'm still having two issues though.

I've placed the following code under 'methods' (not sure if this is the correct way of implementation):

`methods: {
       onStep() {
  
          var prefix = 'question_'    
  
          localStorage.setItem(prefix + this.questions[1].id, this.questions[1].answer)
          localStorage.setItem(prefix + this.questions[2].id, this.questions[2].answer)
          localStorage.setItem(prefix + this.questions[3].id, this.questions[3].answer)
          localStorage.setItem(prefix + this.questions[4].id, this.questions[4].answer)
  
      },
      onKeyListener($event) {
              // We've overriden the default "complete" slot so
              // we need to implement the "keyup" listener manually.
      
              if ($event.key === 'Enter' && this.completed && !this.submitted) {
                this.onSendData()
              }
            }`

The issues that i'm having are:
- localStorage gets set for all the questions above right away as "null", so then on the getItem call the answers are populated with "null" until you type in something else. I need this to be *blank* (if that makes sense)
- The other issue is that i'd like to set the localStorage items one at a time (if possible) as opposed to all getting set right after clicking enter on the first "slide"

**Reply by spinn** [reply]
You could use the `onAnswer` event and save only the last answered question, eg.:

```js
...

onAnswer(question) {
  localStorage.setItem('question_' + question.id, question.answer)
}

...
```

This should solve both your issues since you'll only get answered questions (so no `null` or blank values).

Documentation:
https://www.ditdot.hr/en/docs/vue-flow-form-guide#component-events

**Reply by yuvi100** [reply]
I had to modify from **question.id** (which returned "undefined") to **question.question.id** (which returns the question id). **question.answer** worked to retrieve the value.


I have two remaining issues:
- One that I forgot to mention initially is that answer: localStorage.getItem('...') on a question that has a type: QuestionType.**MultipleChoice**, won't populate with the answer/s saved.

- The second issue is that if you partially fill out the form and you refresh the page, then go through your answers again (now that they populate from localStorage values)... on each step (slide / question) you go over and click 'enter' or 'next', localStorage will set the answer to 'null' even though it should be the answer that has populated already.

**Reply by spinn** [reply]
QuestionType.MultipleChoice returns an array when `multiple` is `true` so you'd need to serialize/unseralize it when storing and loading values from local storage. 

As for `question.id` being undefined - you're completely right, I forgot that we return the question component and not QuestionModel which would probably be more logical. We're planning to publish a new version shortly so I'll make sure to change this behavior. In the meantime I believe you could use `question.question.id`, although this seems a bit inelegant.

**Reply by yuvi100** [reply]
@spinn re-check my comment - deleted the one you responded to whilst you were responding, sorry!

**Reply by yuvi100** [reply]
@spinn i've serialized / unserialized the array values, however the options aren't populating, any idea?

```
new QuestionModel({
            id: 'age',
            title: 'Age',
            type: QuestionType.MultipleChoice,
            multiple: true,
            required: true,
            answer: JSON.parse(localStorage.getItem('question_age')),
            options: [
              new ChoiceOption({
                label: '18 - 24'
              }),
              new ChoiceOption({
                label: '25 - 34'
              }),
              new ChoiceOption({
                label: '35 - 44'
              }),
              new ChoiceOption({
                label: '45 - 60'
              }),
              new ChoiceOption({
                label: '60+'
              })
            ]
})

onAnswer(question) {
        if(question.question.options.length != 0){
          localStorage.setItem('question_' + question.question.id, JSON.stringify(question.answer))
        } else {
          localStorage.setItem('question_' + question.question.id, question.answer)
        }
}

console.log(JSON.parse(localStorage.getItem('question_age')))  //returns ["18 - 24", "25 - 34"]
```

The second issue (which I mentioned in my previous comment is that if you partially fill out the form and you refresh the page, then go through your answers again (now that they populate from localStorage values)... on each step (slide / question) you go over and click 'enter' or 'next', localStorage will set the answer to 'null' even though it should be the answer that has populated already.

Sorry for the hassle - I really appreciate all your help!

**Reply by spinn** [reply]
The issues you're having were actually in Vue Flow Form and not in your implementation. I've now fixed these issues, it should work if you pull the new version from master, but we're also planning to push a new version to npm within a few days.

**Reply by yuvi100** [reply]
Thanks for the update!

Everything is being saved to localStorage and not populating up as 'null' anymore, which is great.

There's still a slight problem:
If someone selects multiple answers e.g. serialised array = ["18 - 24","25 - 34"] then refreshes the form (the values populate correctly) and then clicks enter again to proceed to the next step, the array gets added to the initial serialised array, the localStorage value ends up looking like this: ["18 - 24","25 - 34","18 - 24","25 - 34"] (basically a repeat).

This also means that if someone was to go back over their answers, deselect one of the multiple selected answers, it would still populate.

**Reply by spinn** [reply]
No problem at all! The issue you reported has also been fixed with the latest commit.

**Reply by yuvi100** [reply]
Everything works! thank you!

**Reply by spinn** [reply]
Just a quick update - there was another issue where the `other` option wouldn't get pre-populated, this has now also been fixed. The `onAnswer` event now also sends `QuestionModel` as its argument (before it sent the Question component). This means that you'll need to update your `onAnswer` listener so it accesses QuestionModel directly, eg.:

```js
onAnswer(question) {
  localStorage.setItem('question_' + question.id, JSON.stringify(question.answer))
}
```
